aether-kernel/aether/kernel/api/models.py

```diff
@@ -273,21 +275,23 @@ class Mapping(ExportModelOperationsMixin('kernel_mapping'), TimeStampedModel):
     name = models.CharField(max_length=50, null=False, unique=True)
     definition = JSONField(blank=False, null=False)
     mappingset = models.ForeignKey(to=MappingSet, on_delete=models.CASCADE)
-    projectschemas = models.ManyToManyField(to=ProjectSchema)
+    projectschemas = models.ManyToManyField(to=ProjectSchema, blank=True)
     is_active = models.BooleanField(default=True)
     is_read_only = models.BooleanField(default=False)
 
     # redundant but speed up queries
     project = models.ForeignKey(to=Project, on_delete=models.CASCADE, blank=True, null=True)
 
-    def save(self, **kwargs):
+    def save(self, *args, **kwargs):
         self.project = self.mappingset.project
+        super(Mapping, self).save(*args, **kwargs)
         entities = self.definition.get('entities', {})
-        self.projectschemas.clear()
+        ps_list = []
         for entity_pk in entities.values():
-            self.projectschemas.add(ProjectSchema.objects.get(pk=entity_pk, project=self.project))
-
-        super(Mapping, self).save(**kwargs)
+            ps = ProjectSchema.objects.get(pk=entity_pk)
+            if ps:
+                ps_list.append(ps)
+        self.projectschemas.add(*ps_list)
```

Comment:
:flushed: 

If you move up `super(Mapping, self).save(*args, **kwargs)` next changes are not going to be saved, are they?

On the other hand, are you removing the useless `project_schema` entries?

This `.get(pk=entity_pk, project=self.project))` is needed to ensure that you are not using project schema entries that belong to other projects.

Note: the method `get` will raise an error `ObjectDoesNotExist` if it does not exist, you don't need the `if ps:` condition. :bomb:

Reply:
`object.add()` is a commit action to the db. Save() is moved up because the mapping has to be saved with an id generated before you can add a m2m object. (here)[https://docs.djangoproject.com/en/2.1/topics/db/examples/many_to_many/]. The only issue with this is when using the rest-framework form, project-schemas must be explicitly selected from the list

Reply:
This resolves my first comment, what about the rest? :thinking: 

To avoid problem with forms you need the `blank=True` option, right!

Reply:
Not exactly, in the case of forms, all previous m2m entries even after saved manually are cleared and only those selected using the form are added before a commit. 
https://stackoverflow.com/a/1925784/4032647